File: src/dpl_pipeline/classify.py
"""Provision-level classification: run every rule over every provision.

Deterministic by construction: rules are pre-sorted by id, provisions are
processed in document order, and regex matching has no state. The same
text + the same rule file always produces byte-identical matches.
"""

from __future__ import annotations

import re

from .models import Provision, RuleMatch
from .rules import Rule

_EXCERPT_RADIUS = 140


def classify_provisions(provisions: list[Provision], rules: list[Rule]) -> list[RuleMatch]:
    matches: list[RuleMatch] = []
    for prov in provisions:
        for rule in rules:  # already sorted by rule_id
            m = rule.match(prov.text)
            if m:
                matches.append(
                    RuleMatch(
                        rule_id=rule.rule_id,
                        category=rule.category,
                        policy_area=rule.policy_area,
                        provision_id=prov.provision_id,
                        excerpt=_excerpt(prov.text, m),
                    )
                )
    return matches


def _excerpt(text: str, m: re.Match) -> str:
    start = max(0, m.start() - _EXCERPT_RADIUS)
    end = min(len(text), m.end() + _EXCERPT_RADIUS)
    snippet = re.sub(r"\s+", " ", text[start:end]).strip()
    prefix = "…" if start > 0 else ""
    suffix = "…" if end < len(text) else ""
    return f"{prefix}{snippet}{suffix}"

File: lib/classify.ts
/**
 * Provision-level classification: run every rule over every provision.
 *
 * Deterministic by construction: rules are pre-sorted by id, provisions are
 * processed in document order, and regex matching has no state. The same
 * text + the same rule file always produce identical matches.
 */

import { matchRule, type Rule } from "./rules";
import type { Provision, RuleMatch } from "./types";

const EXCERPT_RADIUS = 140;

export function classifyProvisions(provisions: Provision[], rules: Rule[]): RuleMatch[] {
  const matches: RuleMatch[] = [];
  for (const prov of provisions) {
    for (const rule of rules) {
      const m = matchRule(rule, prov.text);
      if (m) {
        matches.push({
          ruleId: rule.id,
          category: rule.category,
          policyArea: rule.policyArea,
          provisionId: prov.provisionId,
          excerpt: excerpt(prov.text, m.index, m.index + m[0].length),
        });
      }
    }
  }
  return matches;
}

function excerpt(text: string, matchStart: number, matchEnd: number): string {
  const start = Math.max(0, matchStart - EXCERPT_RADIUS);
  const end = Math.min(text.length, matchEnd + EXCERPT_RADIUS);
  const snippet = text.slice(start, end).replace(/\s+/g, " ").trim();
  return `${start > 0 ? "…" : ""}${snippet}${end < text.length ? "…" : ""}`;
}
